ansible_collections/arista/avd/plugins/plugin_utils/eos_validate_state_utils/csv_report.py

```diff
@@ -0,0 +1,64 @@
+# Copyright (c) 2023 Arista Networks, Inc.
+# Use of this source code is governed by the Apache License 2.0
+# that can be found in the LICENSE file.
+from __future__ import annotations
+
+import csv
+from typing import TYPE_CHECKING, Generator
+
+if TYPE_CHECKING:
+    from io import TextIOWrapper
+
+    from .results_manager import ResultsManager
+
+
+class CSVReport:
+    """Generate and writes the validation CSV report based on test results managed by ResultsManager."""
+
+    def __init__(self, csvfile: TextIOWrapper, results: ResultsManager) -> None:
+        """Initialize the CSVReport with an open CSV file object to write to and a ResultsManager instance.
+
+        Args:
+        ----
+            csvfile (TextIOWrapper): An open file object to write the CSV data into.
+            results (ResultsManager): The ResultsManager instance containing all test results.
+        """
+        self.csvfile = csvfile
+        self.results = results
+
+    def generate_rows(self) -> Generator[dict, None, None]:
+        """Generate rows of test result data for the CSV report.
+
+        Results are sourced from `failed_tests` or `all_tests`, based on whether the report includes only failed tests or all results.
+
+        Yields
+        ------
+            Generator[dict, None, None]: A generator of test result dictionaries representing CSV rows.
+        """
+        results = self.results.failed_tests if self.results.only_failed_tests else self.results.all_tests
+
+        for result in results:
+            yield {
+                **result,
+                "failure_reasons": ";".join(result["failure_reasons"]),
+                "test_categories": ";".join(result["test_categories"]),
+            }
+
+    def generate_report(self) -> None:
+        """Generate and writes the CSV report using the collected test results."""
+        fieldnames = ["test_id", "node", "test_categories", "test_description", "test", "result", "failure_reasons"]
+        writer = csv.DictWriter(self.csvfile, fieldnames=fieldnames)
+        writer.writeheader()
+
+        # writerows seems to perfom better when writing in batches
+        batch_size = 5000
+        batch = []
+
+        for row in self.generate_rows():
```

Comment:
probably can use `isslice` here  https://docs.python.org/3/library/itertools.html#itertools.islice
https://stackoverflow.com/questions/6335839/python-how-to-read-n-number-of-lines-at-a-time

Reply:
Outdated